### PR TITLE
GAWB-678: persist timestamps for submission/workflow status changes

### PR DESCRIPTION
**Pull request by davidangb**
Create new tables to record status changes and timestamps for workflows and submissions.

I've done this using database triggers, as opposed to Scala code ... keeps things simple, at the potential cost of splitting our implementation logic into multiple locations.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    <changeSet logicalFilePath="dummy" author="davidan" id="status_change_log">
+
+        <createTable tableName="STAT_WORKFLOW_STATUS">
+            <column autoIncrement="true" name="id" type="BIGINT">
+                <constraints primaryKey="true"/>
+            </column>
+            <column name="workflow_id" type="INT">
+                <constraints nullable="false"/>
+            </column>
```
Comment (davidangb) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml:12]: I am intentionally NOT making these foreign keys. As submissions/workflows are deleted, I do NOT want these deleted; it's worth keeping history over time even for deleted submissions/workflows. Happy to hear objections to this.

Yes, I realize that we would not be able to retrieve method, entity, user, etc. info for any deleted submissions/workflows; we can still keep the status change timing.

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                        (workflow_id, status, timestamp)
+                    VALUES
+                        (NEW.ID, NEW.STATUS, NEW.STATUS_LAST_CHANGED);
+
+            CREATE TRIGGER after_submission_insert AFTER INSERT ON SUBMISSION
+                FOR EACH ROW
+                    INSERT INTO STAT_SUBMISSION_STATUS
+                        (submission_id, status, timestamp)
+                    VALUES
+                        (NEW.ID, NEW.STATUS, CURRENT_TIMESTAMP);
```
Comment (davidangb) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml:74]: for workflows, we can use the timestamp provided by Scala. For submissions, we don't have one of these, so we rely on the database's timestamp. I waffled on this: is it better to be consistent between the two STAT_\* tables, always using db's timestamp? Or is it better to be consistent between WORKFLOW and STAT_WORKFLOW_STATUS, using the exact same value for each?

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="davidan" id="status_change_log">
+
+        <createTable tableName="STAT_WORKFLOW_STATUS">
```
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml]: what does the STAT prefix mean?

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml (hunk 1 above)
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml:12]: 👍

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml (hunk 3 above)
Comment (davidangb) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml]: "statistics" ... not at all attached to this.

**Comment by dvoet** [comment]
I would still lobby for recording the original status as well as the new status. It makes finding unexpected transitions easier.

**Comment by davidangb** [comment]
I thought about persisting the original status and decided against it. The original status is easily figure-out-able just by querying the table, ordered by timestamp, for a given workflow/submission ... and recording it adds to the size of the tables. Still worth it?

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml (hunk 2 above)
Comment (helgridly) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml:74]: I really want to answer "why not both?", but I'll restrain myself.

**Comment by helgridly** [comment]
I'm not a huge fan of database triggers because I don't like hiding business logic in the database, but I'm not going to fight about it. :+1:

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml (hunk 3 above)
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160512_status_change_log.xml]: I propose AUDIT instead, but take it or leave it. Rationale being that this is an audit log and statistics represents only 1 usage of these tables.

**Comment by dvoet** [comment]
one more minor comment but do as you see fit 👍

**Comment by helgridly** [comment]
:+1:

**Comment by dvoet** [comment]
👍

**Pull request merged**